### PR TITLE
CXF-8706: CXF MTOM handler allow content injection

### PR DESCRIPTION
**Pull request by reta**
CXF MTOM handler allow content injection

### CODE REVIEW

**Comment by reta** [comment]
@deki wondering if you have time to check this one out, thank you!

**Comment by dkulp** [comment]
I don't think this is the "right" fix for this.   The Soap w/ Attachement spec allows for http URL's for the cid and actually shows that use in the examples ( https://www.w3.org/TR/SOAP-attachments/ ).  Thus, I think the better fix would be to pass a flag into LazyDataSource to allow flipping to URLDataSource if the CID isn't present as a direct attachment.

**Comment by reta** [comment]
> Thus, I think the better fix would be to pass a flag into LazyDataSource to allow flipping to URLDataSource if the CID isn't present as a direct attachment.

@dkulp thanks, I think I got your idea, but it solves a bit different problem: it does not help with the original issue, namely to limit href scheme `cid:` usage for `xop:include`, unless I am missing something ...

**Comment by dkulp** [comment]
The main issue is that SOAP w/ Attachments doesn't have that limitation.  Thus, applying that limitation in AttachmentUtils is not the right place unless an extra flag is passed in to enable/disable the check and that's just set for the xop cases.

**Comment by reta** [comment]
> The main issue is that SOAP w/ Attachments doesn't have that limitation. Thus, applying that limitation in AttachmentUtils is not the right place unless an extra flag is passed in to enable/disable the check and that's just set for the xop cases.

Exactly, I am looking at how to distinguish xop / non-xop cases, this information is "lost" when MTOM decorator calls `AttachmentUnmarshaller::getAttachmentAsDataHandler` (if you have something in mind, please share), thanks!

**Comment by reta** [comment]
@dkulp relied on `isXOPPackage()` flag of the `AttachmentUnmarshaller` to distinguish xop / non-xop cases, what do you think of it? Thank you!

File: core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java [hunk 1]
```diff
@@ -546,24 +546,40 @@ public static Attachment createMtomAttachmentFromDH(
...
+            // See please https://www.w3.org/TR/xop10/
+            //
+            if (contentId.indexOf("://") == -1) {
+                return loadDataSource(contentId, atts);
+            } else {
+                try {
+                    return new URLDataSource(new URL(contentId));
+                } catch (MalformedURLException e) {
+                    throw new Fault(e);
+                }
```
Comment (coheigea) [inline core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java:578]: Hi Andriy, let's control this behavior with a system property, where it only allows to read message attachments by default.

**Review by coheigea — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java [hunk 2]
```diff
@@ -546,24 +550,46 @@ public static Attachment createMtomAttachmentFromDH(
...
+            // 
+            // See please https://www.w3.org/TR/xop10/
+            //
+            if (contentId.indexOf("://") == -1) {
+                return loadDataSource(contentId, atts);
+            } else {
+                try {
+                    final String followUrls = SystemPropertyAction
+                        .getProperty(ATTACHMENT_XOP_FOLLOW_URLS_PROPERTY, "false");
+                    if ("true".equals(followUrls)) {
```
Comment (coheigea) [inline core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java:581]: Maybe use Boolean.valueOf(followUrls) instead to allow for someone setting "True"?

**Review by coheigea — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java (hunk 2 above)
Comment (reta) [reply core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java:581]: Certainly, thanks @coheigea , changed

Comment (coheigea) [reply core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java:581]: Did you push the update @reta ?
Also this test is failing for me:

[ERROR] Failures:
[ERROR]   MtomServerTest.testURLBasedAttachment:196 expected:<1> but was:<0>
[INFO]

Comment (reta) [reply core/src/main/java/org/apache/cxf/attachment/AttachmentUtil.java:581]: Oops, sorry, just did @coheigea (was fixing exactly this test)

**Review by coheigea — APPROVED** [review]
LGTM

**Pull request merged**